### PR TITLE
feat: add technical specification for balance fetcher component

### PR DESCRIPTION
**Pull request by yahgwai**
## Summary
- Add comprehensive technical specification for the balance fetcher component
- Define clear requirements, API, data structures, and implementation details
- Specify explicit out-of-scope items to maintain focus on core functionality

no-issue: this is a technical specification document

## Test plan
- [ ] Review specification against requirements in ARCHITECTURE.md and REQUIREMENTS.md
- [ ] Verify all component responsibilities are covered
- [ ] Confirm data structures match file-manager.md specification
- [ ] Check that incremental processing requirements are addressed

### CODE REVIEW

File: docs/specs/balance-fetcher.md [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+
+## 1. Overview
+
+The Balance Fetcher is a critical component in the AEP Fee Calculator system that retrieves and stores historical balance information for reward distributors on the Arbitrum Nova network. It queries blockchain state at specific end-of-day blocks to capture point-in-time balance snapshots necessary for accurate fee calculations.
+
+## 2. Purpose
+
+### Primary Objectives
+
+- Retrieve end-of-day ETH balances for all tracked reward distributors
```
Comment (yahgwai) [inline docs/specs/balance-fetcher.md:11]: No, only does it for one distributor - a new instance would be created per distributor

File: docs/specs/balance-fetcher.md [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+- Store historical balance data in a consistent, accessible format
```
Comment (yahgwai) [inline docs/specs/balance-fetcher.md]: Uses the filemanager to do this

File: docs/specs/balance-fetcher.md [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+- Support incremental processing to handle new dates efficiently
+- Handle distributors created mid-period by only fetching balances from their creation date onward
```
Comment (yahgwai) [inline docs/specs/balance-fetcher.md:14]: Fetch balance at the distributor creation time as well

File: docs/specs/balance-fetcher.md [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+
+### Component Responsibilities
+
+- Query blockchain state using RPC calls to retrieve balances
+- Manage balance data persistence for each distributor
```
Comment (yahgwai) [inline docs/specs/balance-fetcher.md:19]: Only for 1 distributor

File: docs/specs/balance-fetcher.md [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+- Ensure data integrity and handle RPC failures gracefully
+- Maintain idempotent operations for reliability
+
+### Success Criteria
+
+- All tracked distributors have balance data for all applicable dates
```
Comment (yahgwai) [inline docs/specs/balance-fetcher.md:25]: Update everywhere in this doc to specify that we're only doing this for one distributor

File: docs/specs/balance-fetcher.md [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+
+- Reliable RPC endpoint with support for `eth_getBalance` at historical blocks
+- File system access for reading distributor list and block numbers
+- Write permissions for storing balance data
+
+## 4. Data Flow
+
+### Input Flow
+
+1. Read distributor addresses from `store/distributors.json`
```
Comment (yahgwai) [inline docs/specs/balance-fetcher.md]: All of this input section is done by calling the file manager

File: docs/specs/balance-fetcher.md [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+   - Load distributor metadata to get creation date
+   - Filter block numbers to only include dates from creation onward
+   - For each applicable date without existing balance data:
+     - Query balance using `eth_getBalance(address, blockNumber)`
+     - Store result in memory
+   - Write updated balance data to distributor's balance file
+
+### Output Flow
+
+- Balance data written to `store/distributors/{address}/balances.json`
```
Comment (yahgwai) [inline docs/specs/balance-fetcher.md]: All updates to file should be done via the file manager - update this doc everyhwere

File: docs/specs/balance-fetcher.md [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+- Each file contains complete balance history for that distributor
+- File updates are atomic (write to temp file, then rename)
+
+## 5. Public API
+
+### Main Entry Point
+
+```typescript
+interface BalanceFetcher {
+  fetchBalances(options?: FetchOptions): Promise<FetchResult>;
```
Comment (yahgwai) [inline docs/specs/balance-fetcher.md]: Lets not offer the fetchoptions at all

File: docs/specs/balance-fetcher.md [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+     - Convert balance from hex to decimal string
+     - Store in balance map
+   - Merge new balances with existing data
+   - Write updated balance file atomically
+
+### Error Handling
+
+- RPC failures trigger exponential backoff retry (up to 3 attempts)
+- Individual distributor failures don't stop processing of others
+- Failed fetches are logged and reported in result summary
```
Comment (yahgwai) [inline docs/specs/balance-fetcher.md]: If failing after retry we should throw error - not log and swallow

File: docs/specs/balance-fetcher.md [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+- Partial data is never written - updates are atomic
```
Comment (yahgwai) [inline docs/specs/balance-fetcher.md]: The file manager handles all this

File: docs/specs/balance-fetcher.md [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+3. **Block Validation**:
+   - Block numbers must be positive integers
+   - Blocks must be at least 1000 blocks old (reorg protection)
+
+### Security Considerations
+
+1. **RPC Security**:
+
+   - Use HTTPS endpoints only
+   - Implement request rate limiting
```
Comment (yahgwai) [inline docs/specs/balance-fetcher.md]: Dont implement this

File: docs/specs/balance-fetcher.md [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+   - Handle RPC errors without exposing sensitive information
+
+2. **Data Integrity**:
```
Comment (yahgwai) [inline docs/specs/balance-fetcher.md:202]: we dont need this section

File: docs/specs/balance-fetcher.md [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+   - Atomic file writes to prevent corruption
+   - Validate all RPC responses before storage
+   - Maintain backup of existing data before updates
+
+## 9. Test Data Requirements
+
+### Minimum Test Data
+
+1. **Distributor List**: At least 3 distributor addresses with different creation dates
```
Comment (yahgwai) [inline docs/specs/balance-fetcher.md]: This can be found in the __tests__/test-data already I believe - note that here if so

File: docs/specs/balance-fetcher.md [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+2. **Block Numbers**: At least 7 days of block numbers spanning distributor lifetimes
```
Comment (yahgwai) [inline docs/specs/balance-fetcher.md]: We have some block numbers in the test-data folder too, check if it's enough and note that here

File: docs/specs/balance-fetcher.md [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+3. **RPC Responses**: Mock responses for `eth_getBalance` calls
```
Comment (yahgwai) [inline docs/specs/balance-fetcher.md]: Dont mock make actual rpc calls

File: docs/specs/balance-fetcher.md [hunk 16]
```diff
@@ -0,0 +1,311 @@
...
+4. **Edge Cases**:
```
Comment (yahgwai) [inline docs/specs/balance-fetcher.md:213]: Forget these lets just use all the distributors from our test data - the ones which are actual reward distributors

File: docs/specs/balance-fetcher.md [hunk 17]
```diff
@@ -0,0 +1,311 @@
...
+
+interface BlockNumbers {
+  [date: string]: number; // Map of ISO date to block number
+}
+```
+
+### Output Types
+
+```typescript
+interface FetchResult {
```
Comment (yahgwai) [inline docs/specs/balance-fetcher.md]: Lets just return void when we're done successfully

File: docs/specs/balance-fetcher.md [hunk 18]
```diff
@@ -68,34 +71,34 @@ The Balance Fetcher is a critical component in the AEP Fee Calculator system tha
...
-  fetchBalances(options?: FetchOptions): Promise<FetchResult>;
+  fetchBalances(): Promise<void>;
 }
 ```
 
 ### Method Signatures
 
 ```typescript
-fetchBalances(options?: FetchOptions): Promise<FetchResult>
+fetchBalances(): Promise<void>
```
Comment (yahgwai) [inline docs/specs/balance-fetcher.md]: Add an optional param to provide a distributor address. If one is provided, filter the distributors to that address.
